eng/packages/http-client-csharp-mgmt/generator/Azure.Generator.Mgmt/src/SystemObjectTypeVisitor.cs

```diff
@@ -0,0 +1,175 @@
+
+using Azure.Generator.Management.Providers;
+using Azure.Generator.Mgmt.Primitives;
+using Microsoft.TypeSpec.Generator.ClientModel;
+using Microsoft.TypeSpec.Generator.ClientModel.Providers;
+using Microsoft.TypeSpec.Generator.Expressions;
+using Microsoft.TypeSpec.Generator.Input;
+using Microsoft.TypeSpec.Generator.Primitives;
+using Microsoft.TypeSpec.Generator.Providers;
+using Microsoft.TypeSpec.Generator.Snippets;
+using Microsoft.TypeSpec.Generator.Statements;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+
+namespace Azure.Generator.Management
+{
+    internal class SystemObjectTypeVisitor : ScmLibraryVisitor
+    {
+        protected override ModelProvider? PreVisitModel(InputModelType model, ModelProvider? type)
+        {
+            if (type is InheritableSystemObjectModelProvider systemType)
+            {
+                UpdateNamespace(systemType);
+            }
+
+            if (type is not InheritableSystemObjectModelProvider && type?.BaseModelProvider is InheritableSystemObjectModelProvider baseSystemType)
+            {
+                Update(baseSystemType, type);
+            }
+            return type;
+        }
+
+        protected override TypeProvider? VisitType(TypeProvider type)
+        {
+            if (type is ModelFactoryProvider modelFactory)
+            {
+                UpdateModelFactory(modelFactory);
+            }
+
+            if (type is InheritableSystemObjectModelProvider systemType)
+            {
+                UpdateNamespace(systemType);
+            }
+
+            if (type is ModelProvider model && model is not InheritableSystemObjectModelProvider && model.BaseModelProvider is InheritableSystemObjectModelProvider baseSystemType)
+            {
+                Update(baseSystemType, model);
+            }
+            return type;
+        }
+
+        // Only this is needed for the current visitor
+        // TODO: The rest of the visitor is temparary workaround for now, once we can generate Azure.ResourceManager from MPG generator, we can remove them
+        private static void UpdateModelFactory(ModelFactoryProvider modelFactory)
+        {
+            var methods = new List<MethodProvider>();
+            foreach (var method in modelFactory.Methods)
+            {
+                var returnType = method.Signature.ReturnType;
+                if (returnType is not null && KnownManagementTypes.IsKnownManagementType(returnType))
+                {
+                    continue;
+                }
+                methods.Add(method);
+            }
+            modelFactory.Update(methods: methods);
+        }
+
+        private static void UpdateNamespace(InheritableSystemObjectModelProvider systemType)
+        {
+            // This is needed because we updated the namespace with NamespaceVisitor in Azure generator earlier
+            systemType.Type.Update(@namespace: systemType._type.Namespace);
```

Comment:
well this might be one of the reasons that we should prompt this type into the dependencies, even MTG itself.
Other part of the generator should know that this provider is not changable.

But since we already write it here, I am fine with the implementation here.
FYI @JoshLove-msft

Reply:
Yeah that is fair. I am fine with moving it to Azure.Generator, but it is also fine to leave it here.